Allow payment requests for same origin-domain subdocuments

zcorpan opened the pull request:
Fixes #2184.

---

This is on top of https://github.com/whatwg/html/pull/2195.

Same origin-domain is used instead of same-origin to disallow by default if `document.domain` had been used in the parent document.

The origin check is part of the snapshot instead of live.

Browsing context containers other than `iframe` (e.g. `object`, `embed`, `frame`) that are same origin-domain are allowed to use payment request.

(I think ideally this should apply to `allowfullscreen` and `allowusermedia` as well, but I don't want to hold up `allowpaymentrequest` on compat concerns for `allowfullscreen` so focusing on just `allowpaymentrequest` right now.)

Please use #2184 for comments about the general approach/concerns, and comment here for editorial review.

---

Implementer sign-off:
* [x] Chromium: @rsolomakhin ? @clelland ?
* [x] Gecko: @bzbarsky ?
* [ ] WebKit: @hober ?
* [ ] Edge: @adrianba ?

annevk commented on source:
```diff
@@ -28615,14 +28615,65 @@ interface <dfn>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
+
+      <li><var>document</var>'s <span data-x="concept-document-bc">browsing context</span> has a
+      <span>browsing context container</span> whose <span>node document</span>'s <span>origin</span>
+      has the <span>same origin-domain</span> as <var>document</var>'s <span>origin</span></li>
```
Why not use same origin here? I think we should keep same origin-domain for legacy cases only.

zcorpan replied:
See the issue for discussion with @bzbarsky. In particular using same-origin would allow cases where script DOM access is otherwise not allowed.

annevk replied:
Right, similar to how access to the network or storage would remain "shared", but script access would not. I guess we could go either way... Or we could require both meaning that when you use `document.domain` you're out of luck with this feature, but that might not be compatible either.

zcorpan replied:
Since it is snapshotted at document creation time it will not work if document.domain is used with the proposed spec. But it would work if "same origin" is used. So same origin-domain + snapshot allows fewer cases without the attribute.

annevk replied:
I see, quite subtle!

clelland replied:
I think that "has the same origin-domain as" should probably read "is same origin-domain with" here.

zcorpan replied:
I used the same wording as elsewhere in the spec.

annevk replied:
Probably wrong elsewhere too then, since it's not really a property of origins, but a check you perform on them.

clelland replied:
https://html.spec.whatwg.org/#svg-0:same-origin-domain is the only instance I could find with that wording; all the other cross-referenced uses seem to use some variant of "is same origin-domain with"

zcorpan replied:
OK fixed. Thanks!

zcorpan replied:
I tried to clarify this a bit now.

annevk commented:
LGTM, but it would be good to have tests (as noted by the label), but also some implementer sign-off. I have the feeling we'll keep tweaking these allow* attributes forever otherwise.

bzbarsky commented:
This seems generally ok to me, but have you talked to the feature policy people about the origin-domain aspect of things?

zcorpan commented:
@igrigorik @mikewest @raymeskhoury @ojanvafai @clelland

It would be good if the spec here for `allowpaymentrequest` could be described in terms of Feature Policy. Can you review this and make sure Feature Policy has the exact same behavior (or propose changes for this PR so they are the same)? Also see comments in #2184.

I would be happy to add any hooks to HTML that FP can use if necessary.

clelland commented:
The wording seems good to me; Same origin-domain seems like the right boundary to be using.

I think this behavior matches the way that Feature Policy should work for allowpaymentrequest -- at least as far as the iframe attribute goes. There may be additional complications when we introduce policies via respone headers, but as long as we can amend this at that point, then this should be good right now.

(It would be great to eventually be able to define 'allowed to use' in terms of feature policy, but that spec isn't quite there yet.)

annevk commented:
Thanks, @clelland do you think we should double check with @igrigorik or should we move ahead and merge?

igrigorik commented:
> The wording seems good to me; Same origin-domain seems like the right boundary to be using.

Ditto.

raymeskhoury commented:
Sorry for being slow to the thread and if this is a silly question, but I'm unsure why we want to check for same origin-domain as opposed to same-origin?

annevk commented:
@raymeskhoury it was discussed in https://github.com/whatwg/html/issues/2184#issuecomment-268911463. Since the check is made during document creation using same origin-domain will allow slightly less than using same-origin would. In particular if the parent had already set document.domain it would not be treated as same-origin.

I think the argument was made that this would be slightly better since the documents can't access each other through script in that case either.

raymeskhoury commented:
Ah so using same origin-domain would be less permissive in this case?

I was just looking at the definition of same origin-domain here: https://w3c.github.io/html/browsers.html#same-origin-domain and it seemed to indicate that origins with different ports would compare equal, which would be more permissive. Will that apply in this case? Is that what we want?

annevk commented:
It would be less permissive because in this case the child global cannot have `document.domain` set yet. Therefore if the parent global did have `document.domain` set, its origin's domain would be non-null and it could never be same-origin domain with its child. And if neither has `document.domain` set, their origins' domain would be null and therefore you'd have to do a normal same origin check which does take the port into account.

(Also, you want https://html.spec.whatwg.org/multipage/browsers.html#same-origin-domain but it happens they're the same in this case I think.)

domenic commented:
IMO we should include a note outlining all the less-permissive-vs.-more-permissive things we've discovered here. Maybe with an example or three.

annevk commented:
There is a note explaining it.

domenic commented:
Yes, I meant explicitly stating that using same origin-domain is less permissive than using same-origin, and perhaps giving an example of where that's the case.

clelland commented:
It's only less permissive in the snapshotting case, right? In the general case, we can't say anything so definite, since it could either restrict communication between documents, or allow it when it wasn't otherwise possible.

(And, like @raymeskhoury, I originally thought that same origin-domain completely ignored ports, I had to re-read the wording carefully a couple of times to convince myself that it wasn't the case. An example to that end would be useful as well)

annevk commented:
We're only discussing snapshotting here. In general same origin-domain should be avoided.

raymeskhoury commented:
@annevk thanks for clarifying! Sorry for my lack of understanding. SGTM then. Any ways we can make it more obvious would be great though :)

zcorpan commented:
OK added a new commit here with examples, and also a new PR with an example for ports at https://github.com/whatwg/html/pull/2257

zcorpan commented:
Sweet, still need to add tests. Plan in https://github.com/w3c/web-platform-tests/issues/4526 (might not need all of these to merge this, but some of them at least)

domenic commented:
I agree with @foolip's comment that we should also switch allowusermedia in this PR if possible.

zcorpan commented:
I've squashed and rebased on master. Since we added `allowusermedia` to the snapshot model in #2195, it is now consistent with `allowpaymentrequest` for the same origin-domain thing as well. I would appreciate a new editorial review to make sure I didn't make a mistake here.

A first set of tests for `allowpaymentrequest` for this is in https://github.com/w3c/web-platform-tests/issues/4615

annevk commented on source:
```diff
@@ -28717,6 +28717,65 @@ interface <dfn>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
     <var>iframe</var> be that element. Otherwise, abort these steps.</p></li>
 
+    <li>
+
```
This newline can be removed.

zcorpan commented:
I had missed to `git add` a fix I meant to have included (12a49c36c2e5d0fe7c2ac5719b8821df868f3539).

zcorpan commented:
Bugs:

https://bugs.chromium.org/p/chromium/issues/detail?id=682280
https://bugzilla.mozilla.org/show_bug.cgi?id=1318988
https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/10656915/
https://bugs.webkit.org/show_bug.cgi?id=167417

zcorpan commented:
I think it is OK to merge this now.

annevk commented:
I think so too. We got tests, bugs, and solid text. The only nitpick I have is that we should maybe also have allowusermedia tests, but implementers should still be writing tests too.

zcorpan commented:
Filed https://github.com/w3c/web-platform-tests/issues/4625 for testing `allowusermedia`.
